oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/web/authentication/logout/OidcBackChannelLogoutFilter.java

```diff
@@ -0,0 +1,183 @@
+import org.springframework.util.Assert;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+/**
+ * A filter for the Client-side OIDC Back-Channel Logout endpoint
+ *
+ * @author Josh Cummings
+ * @since 6.1
+ * @see <a target="_blank" href=
+ * "https://openid.net/specs/openid-connect-backchannel-1_0.html">OIDC Back-Channel Logout
+ * Spec</a>
+ */
+public class OidcBackChannelLogoutFilter extends OncePerRequestFilter {
+
+	private static final String DEFAULT_LOGOUT_URI = "/logout/connect/back-channel/{registrationId}";
+
+	private final Log logger = LogFactory.getLog(getClass());
+
+	private final ClientRegistrationRepository clientRegistrationRepository;
+
+	private final AuthenticationManager authenticationManager;
+
+	private final OAuth2ErrorHttpMessageConverter errorHttpMessageConverter = new OAuth2ErrorHttpMessageConverter();
+
+	private RequestMatcher requestMatcher = new AntPathRequestMatcher(DEFAULT_LOGOUT_URI, "POST");
+
+	private OidcSessionRegistry providerSessionRegistry = new InMemoryOidcSessionRegistry();
+
+	private LogoutHandler logoutHandler = new BackchannelLogoutHandler();
+
+	/**
+	 * Construct an {@link OidcBackChannelLogoutFilter}
+	 * @param clientRegistrationRepository the {@link ClientRegistrationRepository} for
+	 * deriving Logout Token authentication
+	 * @param authenticationManager the {@link AuthenticationManager} for authenticating
+	 * Logout Tokens
+	 */
+	public OidcBackChannelLogoutFilter(ClientRegistrationRepository clientRegistrationRepository,
+			AuthenticationManager authenticationManager) {
+		Assert.notNull(clientRegistrationRepository, "clientRegistrationRepository cannot be null");
+		Assert.notNull(authenticationManager, "authenticationManager cannot be null");
+		this.clientRegistrationRepository = clientRegistrationRepository;
+		this.authenticationManager = authenticationManager;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+			throws ServletException, IOException {
+		RequestMatcher.MatchResult result = this.requestMatcher.matcher(request);
+		if (!result.isMatch()) {
+			chain.doFilter(request, response);
+			return;
+		}
+		String registrationId = result.getVariables().get("registrationId");
+		ClientRegistration registration = this.clientRegistrationRepository.findByRegistrationId(registrationId);
+		if (registration == null) {
+			this.logger.debug("Did not process OIDC Back-Channel Logout since no ClientRegistration was found");
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+		String logoutToken = request.getParameter("logout_token");
+		if (logoutToken == null) {
+			this.logger.debug("Failed to process OIDC Back-Channel Logout since no logout token was found");
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+			return;
+		}
+		OidcLogoutToken token;
+		try {
+			token = authenticate(logoutToken, registration);
```

Comment:
Thanks for the encouragement to move business logic out of the filter. Please see the latest update, which moves the `OidcSessionRegistry` logic into the logout handler. Now there are only three components: `ClientRegistrationRepository`, `AuthenticationManager`, and `LogoutHandler`.

> it would be nice to see most of the logic and work performed by the filter pushed down into the `AuthenticationManager`

I don't really see any other logic fitting in the `AuthenticationManager`. Its job is to authenticate the logout token and it already does that well. I don't think we should have it authenticate and also do other things.

> If the filter only collected the `registrationId`, then `clientRegistrationRepository.findByRegistrationId` could be pushed down.

I'm not so sure about moving `ClientRegistrationRepository`. It seems that the OAuth 2.0 filters are where the tenant is looked up. This is also true in the SAML 2.0 filters (using `RelyingPartyRegistrationRepository`).

Generally speaking, it appears to be an architectural decision to look up the tenant at the filter level, though this decision was made before me and I'm not sure I have all the context about its reasons (one might be to reduce the number of calls to `ClientRegistrationRepository` overall, say if the filters do it, then the result of that query can be passed along to all components throughout the request).

> I think we should also introduce an `AuthenticationConverter `

I'm open to this as a future enhancement in a separate PR. Do you feel it is necessary now? This may be a place to put `ClientRegistrationRepository`.

Reply:
I feel it will likely be a requested enhancement in the future (as other filters where a converter was missing have had requests for this). It's fine by me to wait for that enhancement to be requested first, if that is preferred.